LoRaEngine/modules/LoRaWanNetworkSrvModule/Logger/IotHubLogger.cs

```diff
@@ -0,0 +1,114 @@
+
+        public IotHubLoggerConfiguration Configuration { get; }
+        public LoggerExternalScopeProvider? ScopeProvider { get; }
+
+        public IotHubLoggerProvider(IotHubLoggerConfiguration configuration)
+            : this(configuration, new Lazy<Task<ModuleClient>>(ModuleClient.CreateFromEnvironmentAsync(new[] { new AmqpTransportSettings(TransportType.Amqp_Tcp_Only) })))
+        { }
+
+        internal IotHubLoggerProvider(IotHubLoggerConfiguration configuration, Lazy<Task<ModuleClient>> moduleClientFactory)
+        {
+            Configuration = configuration;
+            ScopeProvider = configuration.UseScopes ? new LoggerExternalScopeProvider() : null;
+            this.moduleClientFactory = moduleClientFactory;
+        }
+
+        public ILogger CreateLogger(string categoryName) =>
+            this.loggers.GetOrAdd(categoryName, n => new IotHubLogger(this, this.moduleClientFactory));
+
+        public void Dispose()
+        {
+            this.loggers.Clear();
+        }
+    }
+
+    internal class IotHubLogger : ILogger
+    {
+        private readonly IotHubLoggerProvider iotHubLoggerProvider;
+        private readonly Lazy<Task<ModuleClient>> moduleClientFactory;
+
+        public IotHubLogger(IotHubLoggerProvider iotHubLoggerProvider,
+                            Lazy<Task<ModuleClient>> moduleClientFactory)
+        {
+            this.iotHubLoggerProvider = iotHubLoggerProvider;
+            this.moduleClientFactory = moduleClientFactory;
+        }
+
+        public IDisposable BeginScope<TState>(TState state) =>
+            this.iotHubLoggerProvider.ScopeProvider?.Push(state) ?? NoopDisposable.Instance;
+
+        public bool IsEnabled(LogLevel logLevel) =>
+            logLevel >= this.iotHubLoggerProvider.Configuration.LogLevel;
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
+        {
+            _ = formatter ?? throw new ArgumentNullException(nameof(formatter));
+
+            if (!IsEnabled(logLevel))
+                return;
+
+            var configuredEventId = this.iotHubLoggerProvider.Configuration.EventId;
+            if (configuredEventId == 0 || configuredEventId == eventId)
+            {
+                _ = Task.Run(async () =>
+                {
+                    try
+                    {
+                        var formattedMessage = LoggerHelper.AddScopeInformation(this.iotHubLoggerProvider.ScopeProvider, formatter(state, exception));
+                        await SendAsync(formattedMessage);
+                    }
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine($"Error when sending log to IoT Hub: {ex}");
+                        throw;
+                    }
+                });
+            }
+        }
+
+        internal virtual async Task SendAsync(string message)
```

Comment:
we are not using a CancellationToken here, as it is fire and forget. I think we could abort any pending requests, when we dispose the Logger (shutting down). Might be too much complexity?

Reply:
I think in any case it might be worth to try to finish sending the log message. I agree that it may not be worth the complexity.

Reply:
Agreed.